feat: Add security/{threat, disclosure, verifying} (#119)
kubewarden/docs: docs/security/threat-model.md

---
sidebar_label: "Threat Model"
title: ""
---

# Threat Model

The [Kubernetes SIG Security team](https://github.com/kubernetes/community/tree/master/sig-security) has
defined an Admission Control Threat Model. The Kubewarden team has evaluated
Kubewarden against this threat model, and strives to provide secure defaults
that cover it. Still, it is recommended for Kubewarden administrators to read
and understand the threat model, and use it to devise their own thread model as
needed.

If more details about each threat is needed, refer to the [SIG-security original document ](https://github.com/kubernetes/sig-security/tree/main/sig-security-docs/papers/admission-control).

## Threat 1 - Attacker floods webhook with traffic preventing its operation

> An attacker who has access to the Webhook endpoint, at the network level, could

Author: indeed! should act as an extended explanation.

timeouts as the  admission controller uses compute power to process the workloads

**Mitigation**
Webhook fail closed and authenticate callers. We are safe, Kubewarden already does that.

## Threat 3 - Attacker exploits misconfiguration of webhook to bypass
An attacker, who has rights to create workloads in the cluster, is able to exploit
a mis-configuration to bypass the intended security control

**Mitigation**
Regular reviews of webhook configuration catch issues.

### Threat 4 - Attacker has rights to delete or modify the k8s webhook  object

An attacker who has Kubernetes API access, has sufficient privileges to delete
the webhook  object in the cluster.

**Mitigation**
RBAC rights are strictly controlled.

**To do**
I think most of the RBAC is not Kubewarden responsibility. But we can help our users if we:
- Warning them in our docs and *suggest* some minimum RBAC to be used.
- Provide a policy which detect RBAC changes and **maybe** block them.

## Threat 5 - Attacker gets access to valid credentials for the webhook
An attacker gains access to valid client credentials for the admission controller webhook

**Mitigation**
Webhook fails closed.

> Kubewarden is failed closed. Thus, we should be fine.

Reviewer: When we say that Kubewarden is failed closed we mean that if, for any reason, Kubewarden stop responding or crashed, the API server will reject the request by default. It does not matter if the request would be accepted by Kubewarden in normal situations. In other words, if Kubewarden stop working for any reason,  the admission request is rejected.

The fail close configuration is a flag in the webhook to tell it what to do if the app it is being called stop responding or take too long to give a response. 

Did I help make it clear? hehe

Kubewarden should implement mutual TLS authentication
We can add in the recommended policies from the `kubewarden-defaults` Helm
chart a policy to drop the `NET_RAW` capability.

### Threat 9 - Attacker steals traffic from the webhook via spoofing
An attacker is able to redirect traffic from the API server which is intended
for the admission  controller webhook by spoofing.

**Mitigation**
Webhook mTLS authentication is used.

**To do**
Kubewarden should implement mutual TLS authentication

### Threat 10 - Abusing a mutation rule to create a privileged container
An attacker is able to cause a mutating admission controller to modify a workload,
such that  it allows for privileged container creation

**Mitigation**
All rules are reviewed and tested.

**To do**
We may came up with some tests to cover this rules reviews
We  should carefully review a PR changing the rules in the policies deployment

## Threat 11 - Attacker deploys workloads to namespaces that are  exempt from admission control
An attacker is able to deploy workloads to Kubernetes namespaces that are exempt
from the  admission controller configuration.

**Mitigation**
RBAC rights are strictly controlled

**To do**

> I think most of the RBAC is not Kubewarden responsability. But we can help our users if we:

Author: Done in https://github.com/kubewarden/docs/pull/119/commits/b0aa443bb8c68ec6ca9d017c505eff10ba291d5d

- Warning them in our docs and *suggest* some minimum RBAC to be used.
- Provide a policy which detect RBAC changes and **maybe** block them. Is this possible?

## Threat ID 12 - Block rule can be bypassed due to missing match (e.g.  missing initcontainers)
An attacker created a workload manifest which uses a feature of the Kubernetes
API which  is not covered by the admission controller

**Mitigation**
All rules are reviewed and tested.

**To do**

> We may came up with some tests to cover this rules reviews

Author: Done in https://github.com/kubewarden/docs/pull/119/commits/b0aa443bb8c68ec6ca9d017c505eff10ba291d5d

We  should carefully review a PR changing the rules in the policies deployment

## Threat ID 13 - Attacker exploits bad string matching on a blocklist to  bypass rules
An attacker, who has rights to create workloads, bypasses a rule by exploiting

> bad string  matching.

Reviewer 2: ```suggestion
bad string matching.
```

**Mitigation**
All rules are reviewed and tested.

**To do**

> We may came up with some tests to cover this rules reviews

Author: Done in https://github.com/kubewarden/docs/pull/119/commits/b0aa443bb8c68ec6ca9d017c505eff10ba291d5d

We  should carefully review a PR changing the rules in the policies deployment

## Threat ID 14 - Attacker uses new/old features of the Kubernetes API  which have no rules

> An attacker, with rights to create workloads, uses new features of the Kubernetes

Reviewer 2: ```suggestion
An attacker, with rights to create workloads uses new features of the Kubernetes
```

Author: I'm going to not apply this one, even if the current thing is too stackato,I personally thing it's clearer.

API (for  example a changed API version) to bypass a rule.

**Mitigation**
All rules are reviewed and tested.

**To do**

> We may came up with some tests to cover this rules reviews and API versions.

Reviewer 2: Avoid usage of first-person single/plural terms

We can create a configuration to reject by default requests where the API
version not cover by the policy.  We should warning policies developers to cover
all the supported API version in theirs tests and reject all of others.

## Threat ID 15 - Attacker deploys privileged container to node running  Webhook controller
An attacker, who has rights to deploy privileged containers to the cluster, creates

> a privileged  container on the cluster node where the admission controller webhook operates.

Reviewer 2: ```suggestion
a privileged container on the cluster node where the admission controller webhook operates.
```

**Mitigation**
Admission controller uses restrictive policies to prevent privileged  workloads

**To do**

> I do not know if Kuberwarden can help on this. Kubewarden does not have access to

Reviewer 2: Avoid usage of first-person single/plural terms

Author: Done in https://github.com/kubewarden/docs/pull/119/commits/b0aa443bb8c68ec6ca9d017c505eff10ba291d5d

containers running in the cluster node.

## Threat ID 16 - Attacker mounts a privileged node hostpath allowing  modification of Webhook controller configuration
An attacker, who has rights to deploy hostPath volumes with workloads, creates a

> volume  which allows for access to the admission controller pod’s files.

Reviewer 2: ```suggestion
volume that allows for access to the admission controller pod’s files.
```

**Mitigation**
Admission controller uses restrictive policies to prevent privileged  workloads

**To do**

> We can add a recommended policy in the `kubewarden-default` Helm chart to prevent this.

Reviewer 2: Avoid usage of first-person single/plural terms

Author: Done in https://github.com/kubewarden/docs/pull/119/commits/b0aa443bb8c68ec6ca9d017c505eff10ba291d5d

## Threat ID 17 - Attacker has privileged SSH access to cluster node  running admission webhook
An attacker is able to log into cluster nodes as a privileged user via SSH.

> **Mitigation**

Reviewer 2: ```suggestion

**Mitigation**
```

N/A

> I don't think Kubewarden can help on this

Reviewer 2: Avoid usage of first-person single/plural terms

## Threat ID 18 - Attacker uses policies to send confidential data from  admission requests to external systems
An attacker is able to configure a policy that listens to admission requests and

> sends  sensitive data to an external system.

Reviewer 2: ```suggestion
sends sensitive data to an external system.
```

**Mitigation**
Strictly control external access for webhook

Kubewarden policies run in a restrictive environment. They do not have network access.

## Threat Kubewarden ID 1 - Bootstrapping of trust for admission controller

> Assuming a trusted but fresh Kubernetes cluster, an attacker is able to compromise the Kubewarden stack before any of the policies securing it is deployed and enforcing. For example, by using unsigned and malicious images for kubewarden-controller, policy-server, or any of the Kubewarden dependencies (cert-manager) or optional dependencies (grafana, prometheus..), or by compromising the Helm charts payload.

Reviewer 2: Wrapping text + few nits
```suggestion
Assuming a trusted but fresh Kubernetes cluster, an attacker is able to compromise the Kubewarden stack 
before any of the policies securing it are deployed and enforced.  For example, by using unsigned and 
malicious images for kubewarden-controller, policy-server, or any of the Kubewarden dependencies 
(cert-manager) or optional dependencies (grafana, prometheus, etc.), or by compromising 
the Helm charts payload.
```

** Mitigation **
1. Kubewarden provides a Software Bill Of Materials, which lists all images needed. This aids with Zero-Trust.

>   The Kubernetes Administrator must verify the Kubewarden images (and its dependencies' images and charts) out of the Kubernetes cluster, in a trusted environment. This can be done with `cosign`, for example.

Reviewer 2: ```suggestion
  The Kubernetes Administrator must verify the Kubewarden images, its dependencies' images, and charts 
  out of the Kubernetes cluster, in a trusted environment. This can be done with `cosign`, for example.
```

  Incidentally, this is part of the implementation needed for air-gapped installations.

>  2. Use signed Helm charts, and verified digests (instead of tags) for Kubewarden images in those Helm charts. This doesn't secure dependencies though.

Reviewer 2: ```suggestion
 2. Use signed Helm charts, and verified digests instead of tags for Kubewarden images in those Helm charts. 
 This doesn't secure dependencies though.
```

**Mitigation**
Webhook fails closed.

Kubewarden default behaviour is failed closed. Thus, it covers this.

(Failing closed means that if, for any reason, Kubewarden stops responding or
crashes, the API server will reject the request by default, even if the request
would be accepted by Kubewarden in normal situations)

## Threat  6 - Attacker gains access to a cluster admin credential

An attacker gains access to a cluster-admin level credential in the kubernetes cluster.

**Mitigation**
N/A

## Threat 7 - Attacker sniffs traffic on the container network
An attacker who has access to the container network is able to sniff traffic
between the API  server and the admission controller webhook.

**Mitigation**
Since the webhook uses TLS encryption for all traffic, Kubewarden is safe.

## Threat 8 - Attacker carries out a MITM attack on the webhook
An attacker on the container network, who has access to the NET_RAW capability
can attempt to use MITM tooling to intercept traffic between the API server
and admission controller webhook.

**Mitigation**
Webhook mTLS authentication is used.

**To do**

> Kubewarden should implement mutual TLS authentication

Reviewer 3: Yes, this is something we might want to do in the future

## Threat ID 12 - Block rule can be bypassed due to missing match (e.g.  missing initcontainers)
An attacker created a workload manifest which uses a feature of the Kubernetes
API which is not covered by the admission controller

**Mitigation**
All rules are reviewed and tested.

**To do**
Introduce tests to cover this rule.
As always, carefully review PRs changing the rules in the policies deployment.

## Threat ID 13 - Attacker exploits bad string matching on a blocklist to  bypass rules
An attacker, who has rights to create workloads, bypasses a rule by exploiting
bad string matching.

**Mitigation**
All rules are reviewed and tested.

**To do**
Introduce tests to cover this rule.
As always, carefully review PRs changing the rules in the policies deployment.

## Threat ID 14 - Attacker uses new/old features of the Kubernetes API  which have no rules
An attacker, with rights to create workloads, uses new features of the Kubernetes
API (for  example a changed API version) to bypass a rule.

**Mitigation**
All rules are reviewed and tested.

**To do**
Introduce tests to cover this rule.
Create a configuration to reject by default requests where the API version not

> cover by the policy. Kubewarden should warn policy developers to cover all the

Reviewer 2: ```suggestion
cover by the policy by default. Kubewarden should warn policy developers to cover all the
```

> supported API version in theirs tests and reject all of others.

Reviewer 2: ```suggestion
supported API version in their tests and reject all others.
```

An attacker, who has rights to deploy hostPath volumes with workloads, creates a
volume that allows for access to the admission controller pod’s files.

**Mitigation**
Admission controller uses restrictive policies to prevent privileged  workloads

**To do**
Add a recommended policy in the `kubewarden-default` Helm chart to prevent this.

## Threat ID 17 - Attacker has privileged SSH access to cluster node  running admission webhook
An attacker is able to log into cluster nodes as a privileged user via SSH.

**Mitigation**
N/A

## Threat ID 18 - Attacker uses policies to send confidential data from  admission requests to external systems
An attacker is able to configure a policy that listens to admission requests and
sends sensitive data to an external system.

**Mitigation**
Strictly control external access for webhook

Kubewarden policies run in a restrictive environment. They do not have network access.

## Threat Kubewarden ID 1 - Bootstrapping of trust for admission controller
Assuming a trusted but fresh Kubernetes cluster, an attacker is able to compromise the Kubewarden stack 

> before any of the policies securing it are deployed and enforced.  For example, by using unsigned and 

Reviewer 2: ```suggestion
before any of the policies securing it are deployed and enforced. For example, by using unsigned and 
```
